### PR TITLE
[TOPI] Add handwritten matvec for dynamic cases

### PR DESCRIPTION
**Pull request by tkonolige**
Add a handwritten matrix-vector multiplication implementation for dynamic cases on cpu. This avoids crashing when a dynamic shape is present. Only supports vector LHS and float32 for now.

### CODE REVIEW

File: python/tvm/relay/op/strategy/x86.py [hunk 1]
```diff
@@ -507,10 +507,26 @@ def matmul_strategy_cpu(attrs, inputs, out_type, target):
     return strategy
 
 
+def is_any(shape):
```
Comment (jwfromm) [inline python/tvm/relay/op/strategy/x86.py:510]: `is_dynamic` might be a more appropriate name for this function.

File: python/tvm/relay/op/strategy/x86.py [hunk 2]
```diff
@@ -507,10 +507,26 @@ def matmul_strategy_cpu(attrs, inputs, out_type, target):
...
+
+
 @dense_strategy.register("cpu")
 def dense_strategy_cpu(attrs, inputs, out_type, target):
     """dense x86 strategy"""
+
     strategy = _op.OpStrategy()
+    # For dynamic shapes we use a hand written kernel. Right now it only
+    # supports matrix-vector multiplication.
+    if is_any(inputs[0].shape) or is_any(inputs[1].shape):
```
Comment (jwfromm) [inline python/tvm/relay/op/strategy/x86.py:521]: should we check the size of `inputs[0]` to make sure its a vector before dispatching?

**Review by jwfromm — APPROVED** [review]
This all makes sense to me, I left a few nitpicks but otherwise should be good to go.

File: python/tvm/relay/op/strategy/x86.py (hunk 2 above)
Comment (tkonolige) [reply python/tvm/relay/op/strategy/x86.py]: Added a check in the implementation.

File: python/tvm/relay/op/strategy/x86.py (hunk 1 above)
Comment (tkonolige) [reply python/tvm/relay/op/strategy/x86.py]: renamed to `is_dynamic_shape`

**Comment by tkonolige** [comment]
@jwfromm a change: the dynamic kernel usage is predicated on it being a matrix-vector multiply instead of asserting within the kernel. Apparently there are some other code paths that do dynamic matmul that I broke.

**Comment by jwfromm** [comment]
Yeah that makes perfect sense to me, good change.

**Comment by tkonolige** [comment]
@tvm-bot rerun

**Comment by michalpiszczek** [comment]
@tvm-bot rerun

**Comment by michalpiszczek** [comment]
@tvm-bot rerun

**Review by junrushao — APPROVED** [review]
LGTM!

**Pull request merged**